### PR TITLE
Small fixes

### LINKED ISSUE #419: attributeError: 'Model' object has no attribute `drift_analog_weights`
## Description

When the model has an `nn.Fallten` layer, the converted model does not have `drift_analog_weights` and `program_analog_weights` methods.

## How to reproduce

```python
import numpy as np
import torch
from aihwkit.inference import PCMLikeNoiseModel
from aihwkit.nn.conversion import convert_to_analog
from aihwkit.simulator.configs import InferenceRPUConfig
from aihwkit.simulator.configs.utils import MappingParameter, WeightNoiseType
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor

# Set Device
device = "cuda" if torch.cuda.is_available() else "cpu"


class Network(nn.Module):
    def __init__(self, **kwargs):
        super(Network, self).__init__()
        self.flatten = nn.Flatten()
        self.linear_relu = nn.Sequential(
            nn.Linear(28 * 28, 100),
            nn.ReLU(),
            nn.Linear(100, 100),
            nn.ReLU(),
            nn.Linear(100, 10),
            nn.Softmax(dim=1),
        )

    def forward(self, input):
        input = self.flatten(input)
        return self.linear_relu(input)

    def test(self, dataloader):
        """
        test the model on the given dataloader
        Args:
            dataloader: dataloader to evaluate the model on
        Returns:
            accuracy: accuracy of the model on the given dataloader in percentage
        """
        self.eval()
        correct = 0
        with torch.no_grad():
            # Iterate over the test data in parallel
            for X, y in dataloader:
                X, y = X.to(device), y.to(device)
                output = self(X)
                pred = output.argmax(dim=1)
                correct += pred.eq(y).sum().item()
        return correct / len(dataloader.dataset) * 100


# Load test data
loaded_data = datasets.MNIST(
    root="data", train=False, download=True, transform=ToTensor()
)
test_data = DataLoader(loaded_data, batch_size=512, shuffle=False, num_workers=4)

# Load model
FP32_model = Network().to(device)
FP32_model.eval()

# Convert model to analog mapped
# Define device and chip configuration used in the RPU tile
mapping = MappingParameter(
    max_input_size=512,  # analog tile size
    max_output_size=512,
    digital_bias=True,
    weight_scaling_omega=1.0,
)  # whether to use analog or digital bias

print(f"Test Accuracy FP32: {FP32_model.test(test_data)}")

rpu_config = InferenceRPUConfig(mapping=mapping)
# specify additional options of the non-idealities in forward to your liking
rpu_config.forward.inp_res = 1 / 256.0  # 6-bit DAC discretization.
rpu_config.forward.out_res = 1 / 256.0  # 8-bit ADC discretization.
rpu_config.forward.w_noise_type = WeightNoiseType.ADDITIVE_CONSTANT
rpu_config.forward.w_noise = 0.02  # Some short-term w-noise.
rpu_config.forward.out_noise = 0.02  # Some output noise.

# specify the noise model to be used for inference only
rpu_config.noise_model = PCMLikeNoiseModel(g_max=25.0)  # the model described

# Convert the model to its analog version.
# this will replace ``Linear`` layers with ``AnalogLinearMapped``
model_analog = convert_to_analog(FP32_model, rpu_config, realistic_read_write=False)
model_analog.program_analog_weights()  # can also omitted as it is called below in any case
acc = model_analog.test(test_data)
print(f"Test Accuracy before drift {rpu_config.__class__.__name__}: {acc:.2f}")

t_inference = 3600.0  # time of inference in seconds (after programming)
model_analog.drift_analog_weights(
    t_inference
)  # modifies weights according to noise model

acc = model_analog.test(test_data)
print(f"Test Accuracy {rpu_config.__class__.__name__}: {acc:.2f}")

```

## Expected behavior

model should convert to analog and apply tye drift
## Other information

* Pytorch version: 1.12.1
* Package version: Built from source
* OS: Ubuntu 22.04
* Python version: 3.10.6
* Conda version (or N/A): N/A

### LINKED ISSUE #446: Missing zero_grad() call
## Description

It seems like the first two examples (`examples/01_simple_layer.py` and `examples/02_multiple_layers.py`) are missing a call to `opt.zero_grad()` at each iteration. When running on 10 iterations currently the loss decrease then increase again:
```
Loss error: 1.0298337936401367
Loss error: 0.8178885579109192
Loss error: 0.4703792631626129
Loss error: 0.1962183117866516
Loss error: 0.0619336031377316
Loss error: 0.0884876772761345
Loss error: 0.2775749266147614
Loss error: 0.5320941805839539
Loss error: 0.7531698346138000
Loss error: 0.8695605397224426
```

## How to reproduce

Run `examples/01_simple_layer.py` or `examples/02_multiple_layers.py` (it is more noticeable on 10 iterations than 100).

## Expected behavior

The loss should decrease consistently, as when `opt.zero_grad()` is added:
```
Loss error: 1.1083202362060547
Loss error: 0.3732085227966309
Loss error: 0.1592179238796234
Loss error: 0.0920740887522697
Loss error: 0.0612552836537361
Loss error: 0.0359693244099617
Loss error: 0.0363772809505463
Loss error: 0.0245949663221836
Loss error: 0.0224329363554716
Loss error: 0.0218768958002329
```

## Other information

simply adding `opt.zero_grad()` after each `opt.step()` should be sufficient to solve this.

### LINKED ISSUE #447: Indice size Error in rpu_onesided_device
## Description

IBM/aihwkit/tree/master/src/rpucuda/rpu_onesided_device.cpp
208 line initUpdateCycle()
a_indices_ ,b_indices_ are resized into too small size
## How to reproduce
If you simulate FCN which has many nodes like 784x256x128x10 with OneSidedUnitCell
Because we need x_size of a_indice and b_indice at max in sparse update

## Expected behavior
Segmentation Fault or Errors related to memory size occur occasionly

## Other information
 213-216 line have to be changed into
if (a_indices_.size() < (size_t)this->x_size_) {
    a_indices_.resize(this->x_size_);
    b_indices_.resize(this->x_size_);
  }

* Pytorch version:
* Package version:
* OS:
* Python version:
* Conda version (or N/A):

### PR DESCRIPTION
**Pull request by maljoras**
## Related issues

closes #447
closes #446
closes #419 

## Description

* fix indexing error 
* add `zero_grad` call in example 02
* changelog update
* added stand-alone functions for applying drift : 

##Details
Apply drift to any model had required the model to be of `AnalogSequential`. Now one can simply use: 

```python
from aihwkit.inference import drift_analog_weights
model.eval()
drift_analog_weights(model, t_inference=100.)
# ... do inference with drifted model
```

### CODE REVIEW

**Review by Fabio-83 — APPROVED** [review]
many thanks @maljoras
